### PR TITLE
Expand spec with all but SharedWorkers

### PR DESCRIPTION
**Pull request by arichiv**
SharedWorker has a more complex implementation than the others, so let's skip that for now and just define the others.

***
<a href="https://pr-preview.s3.amazonaws.com/privacycg/saa-non-cookie-storage/pull/18.html" title="Last updated on Mar 19, 2024, 4:58 PM UTC (805bb4f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/privacycg/saa-non-cookie-storage/18/abc5abe...805bb4f.html" title="Last updated on Mar 19, 2024, 4:58 PM UTC (805bb4f)">Diff</a>

### CODE REVIEW

File: spec.bs [hunk 1]
```diff
@@ -108,7 +167,7 @@ For now {{Document/hasStorageAccess()}} is not considered deprecated, but that <
 When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>requestStorageAccess(types)</code></dfn> method must run these steps:
 
 1. Let |p| be [=a new promise=].
-1. If |types|.{{StorageAccessTypes/all}} is `false` and |types|.{{StorageAccessTypes/cookies}} is `false`:
+1. If |types|.{{StorageAccessTypes/all}} is `false` and |types|.{{StorageAccessTypes/cookies}} is `false` and |types|.{{StorageAccessTypes/sessionStorage}} is `false` and |types|.{{StorageAccessTypes/localStorage}} is `false` and |types|.{{StorageAccessTypes/indexedDB}} is `false` and |types|.{{StorageAccessTypes/locks}} is `false` and |types|.{{StorageAccessTypes/caches}} is `false` and |types|.{{StorageAccessTypes/getDirectory}} is `false` and |types|.{{StorageAccessTypes/estimate}} is `false` and |types|.{{StorageAccessTypes/createObjectURL}} is `false` and |types|.{{StorageAccessTypes/revokeObjectURL}} is `false` and |types|.{{StorageAccessTypes/BroadcastChannel}} is `false`:
```
Comment (johannhof) [inline spec.bs:172]: Ooof

File: spec.bs [hunk 2]
```diff
@@ -134,33 +192,99 @@ Modify {{Document/requestStorageAccess()}} at step 14.1.1.1.1 to read:
 
 <h3 id="storage">Changes to various client-side storage mechanisms</h3>
 
-<h4 id="dom-storage">DOM Storage</h4>
+For all of the following getters and methods, consider the |docForSAA| to be |doc| with the following modifications:
 
-TBD
+1. No existing storage mechanism caches on |doc| will be available on |docForSAA|.
+1. Any storage mechanism caches made on |docForSAA| will not be added to |doc|.
```
Comment (johannhof) [inline spec.bs]: I'm not an expert on the storage spec unfortunately, but it doesn't feel like "storage mechanism cache" is a normative term. Is there something more normative we could use here, and if not, could you file an issue and add a TODO(#issue number) comment to this section?

File: spec.bs [hunk 3]
```diff
@@ -134,33 +192,99 @@ Modify {{Document/requestStorageAccess()}} at step 14.1.1.1.1 to read:
 
 <h3 id="storage">Changes to various client-side storage mechanisms</h3>
 
-<h4 id="dom-storage">DOM Storage</h4>
+For all of the following getters and methods, consider the |docForSAA| to be |doc| with the following modifications:
```
Comment (johannhof) [inline spec.bs]: We should also mention that this assumes a world where https://github.com/whatwg/storage/pull/144 is merged, i.e. double-keyed (+ AC bit) storage exists.

File: spec.bs (hunk 1 above)
Comment (arichiv) [reply spec.bs:172]: image

File: spec.bs (hunk 3 above)
Comment (johannhof) [reply spec.bs]: Sorry, did you do this? Maybe I'm just not seeing it.

File: spec.bs [hunk 4]
```diff
@@ -134,33 +192,99 @@ Modify {{Document/requestStorageAccess()}} at step 14.1.1.1.1 to read:
...
+1. When attempting to [=obtain a storage key=] the returned key will use [[STORAGE-PARTITIONING#relaxing-additional-keying]] if the tuple does not simply contain an [=/origin=].
 
-<h4 id="indexed-db">IndexedDB</h4>
+Issue(19): Clarify client-side storage mechanism changes in more detail.
 
-TBD
+<h4 id="dom-storage">[=Web storage=]</h4>
 
-<h4 id="web-locks">Web Locks</h4>
+When invoked on {{StorageAccessHandle}} |handle| with {{StorageAccessTypes}} |types| and {{Document}} |docForSAA|, the <dfn export attribute for=StorageAccessHandle><code>sessionStorage</code></dfn> getter must run these steps:
```
Comment (johannhof) [inline spec.bs]: I think we need to make it clear how the handle and types are connected, we can't just conjure the types out of nowhere. Maybe, below the definition of StorageAccessHandle, add roughly something like this:

```
A {{StorageAccessHandle}} object has an associated {{StorageAccessTypes}} <dfn for=StorageAccessHandle>types</dfn>.
```

then when creating the handle you could say
```
Set |handle|'s [=StorageAccessHandle/types=] to |types|.
```

and take the types from the handle when using it here and in the instances below.

File: spec.bs (hunk 3 above)
Comment (johannhof) [inline spec.bs]:
```suggestion
For all of the following getters and methods, consider the following modifications:
```

Comment (johannhof) [reply spec.bs]: I'm not sure how the document plays into this, so I would suggest removing it for now but filing a new issue for wiring up a "unpartitioned" bit into the `[=obtain a storage key=]` algorithm, I think that's actually not that hard, we just have to trace it from all the getters monkeypatched below.

In the meantime, I think this is clear enough.

File: spec.bs [hunk 5]
```diff
@@ -134,33 +192,99 @@ Modify {{Document/requestStorageAccess()}} at step 14.1.1.1.1 to read:
...
 
-TBD
+1. If |types|.{{StorageAccessTypes/all}} is `false` and |types|.{{StorageAccessTypes/sessionStorage}} is `false`:
+    1. Throw an "{{InvalidStateError}}" {{DOMException}}.
+1. Return the invocation of [=html/sessionStorage=] on |docForSAA|.
```
Comment (johannhof) [inline spec.bs]: As mentioned above, sessionStorage is not actually invoked on a document, it gets the window's (this) associated document. But I don't think that actually matters so much, the handle holds the information that is sufficient to tell the getter that it needs a different storage key.

So I think we can remove the mention of |docForSAA| here, WDYT?

**Review by johannhof — APPROVED** [review]
LGTM for iterative process (% some comments), let's follow up on a few of these.

**Pull request merged**